pkg/apis/pipeline/v1beta1/pipeline_conversion.go

```diff
@@ -20,23 +20,247 @@ import (
+	}
+	ps.Results = nil
+	for _, r := range source.Results {
+		new := PipelineResult{}
+		new.convertFrom(ctx, r)
+		ps.Results = append(ps.Results, new)
+	}
+	ps.Finally = nil
+	for _, f := range source.Finally {
+		new := PipelineTask{}
+		new.convertFrom(ctx, f)
+		ps.Finally = append(ps.Finally, new)
+	}
+	return nil
+}
+
+func (pt PipelineTask) convertTo(ctx context.Context, sink *v1.PipelineTask) {
+	sink.Name = pt.Name
+	if pt.TaskRef != nil {
+		newTaskRef := &v1.TaskRef{}
+		pt.TaskRef.convertTo(ctx, newTaskRef)
+		sink.TaskRef = newTaskRef
+	}
+	if pt.TaskSpec != nil {
+		newTaskSpec := &v1.EmbeddedTask{}
+		pt.TaskSpec.convertTo(ctx, newTaskSpec)
+		sink.TaskSpec = newTaskSpec
+	}
+	sink.When = nil
+	for _, we := range pt.WhenExpressions {
+		new := v1.WhenExpression{}
+		we.convertTo(ctx, &new)
+		sink.When = append(sink.When, new)
+	}
+	sink.Retries = pt.Retries
+	sink.RunAfter = pt.RunAfter
+	sink.Params = nil
+	for _, p := range pt.Params {
+		new := v1.Param{}
+		p.convertTo(ctx, &new)
+		sink.Params = append(sink.Params, new)
+	}
+	sink.Matrix = nil
+	for _, m := range pt.Matrix {
+		new := v1.Param{}
+		m.convertTo(ctx, &new)
+		sink.Matrix = append(sink.Matrix, new)
 	}
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", source)
+	sink.Workspaces = nil
+	for _, w := range pt.Workspaces {
+		new := v1.WorkspacePipelineTaskBinding{}
+		w.convertTo(ctx, &new)
+		sink.Workspaces = append(sink.Workspaces, new)
+	}
+
+	sink.Timeout = pt.Timeout
+}
+
+func (pt *PipelineTask) convertFrom(ctx context.Context, source v1.PipelineTask) {
+	pt.Name = source.Name
+	if source.TaskRef != nil {
+		newTaskRef := TaskRef{}
+		newTaskRef.convertFrom(ctx, *source.TaskRef)
+		pt.TaskRef = &newTaskRef
+	}
+	if source.TaskSpec != nil {
+		newTaskSpec := EmbeddedTask{}
+		newTaskSpec.convertFrom(ctx, *source.TaskSpec)
+		pt.TaskSpec = &newTaskSpec
+	}
+	pt.WhenExpressions = nil
+	for _, we := range source.When {
+		new := WhenExpression{}
+		new.convertFrom(ctx, we)
+		pt.WhenExpressions = append(pt.WhenExpressions, new)
+	}
+	pt.Retries = source.Retries
+	pt.RunAfter = source.RunAfter
+	pt.Params = nil
+	for _, p := range source.Params {
+		new := Param{}
+		new.convertFrom(ctx, p)
+		pt.Params = append(pt.Params, new)
+	}
+	pt.Matrix = nil
+	for _, m := range source.Matrix {
+		new := Param{}
+		new.convertFrom(ctx, m)
+		pt.Matrix = append(pt.Matrix, new)
+	}
+	pt.Workspaces = nil
+	for _, w := range source.Workspaces {
+		new := WorkspacePipelineTaskBinding{}
+		new.convertFrom(ctx, w)
+		pt.Workspaces = append(pt.Workspaces, new)
+	}
+
+	pt.Timeout = source.Timeout
+}
+
+func (et EmbeddedTask) convertTo(ctx context.Context, sink *v1.EmbeddedTask) error {
+	sink.TypeMeta = et.TypeMeta
+	sink.Spec = et.Spec
+	sink.Metadata = v1.PipelineTaskMetadata(et.Metadata)
+	newTaskSpec := v1.TaskSpec{}
+	err := et.TaskSpec.ConvertTo(ctx, &newTaskSpec)
+	if err != nil {
+		return err
+	}
+	sink.TaskSpec = newTaskSpec
+	return nil
```

Comment:
Can this block be replaced by `return et.TaskSpec.ConvertTo(ctx, sink.TaskSpec)`? (and same question for `EmbeddedTask.convertFrom`)